File: src/input/scene_parser.rs
use crate::geo;
use geo::Point3;
use geo::Transformation;
use geo::Vec3;

use crate::object;
use object::Albedo;
use object::Brdf;
use object::Scene;
use object::Camera;
use object::Color;
use object::Light;
use object::NormalMap;
use object::Primitive;
use object::Radiance;
use object::Shape;
use object::Surface;

use object::albedo::Solid;
use object::albedo::Texture;

use object::brdf::Lambert;
use object::brdf::OrenNayar;
use object::brdf::Phong;
use object::brdf::TorranceSparrow;

use object::shape::Sphere;
use object::shape::Quad;
use object::shape::Transformed;

use crate::input;
use input::BmpLoader;
use input::BptLoader;
use input::PlyLoader;

use std::fs;
use std::cmp::min;

pub struct SceneParser {
    data: String,
    pos: usize,
    line: usize
}

struct ParseError {
    message: String
}

impl ParseError {
    pub fn new(message: &str) -> ParseError {
        return ParseError { message: message.to_string() };
    }
}

impl SceneParser {
    pub fn new(filename: String) -> SceneParser
    {
        let data = fs::read_to_string(filename).unwrap_or_default();
        return SceneParser { data, pos: 0, line: 1 };
    }

    pub fn parse(&mut self) -> Option<Scene> {
        if self.data.is_empty() {
            return None;
        }

        self.skip_whitespace();
        match self.parse_scene() {
            Ok(scene) => return Some(scene),
            Err(err) => {
                println!("Error, line {} : {}", self.line, err.message);
                return None;
            }
        }
    }

    fn skip_whitespace(&mut self) {
        while self.pos < self.data.len() {
            let c = self.data.chars().nth(self.pos).unwrap_or_default();
            if char::is_whitespace(c) || c == '\r' {
                self.pos += 1;
                continue;
            } else if c == '\n' {
                self.line += 1;
                self.pos += 1;
            } else {
                break;
            }
        }
    }

    fn match_end(&mut self) -> bool {
        return self.pos == self.data.len();
    }

    fn match_literal(&mut self, literal: &str) -> bool {
        if self.match_end() {
            return false;
        }

        let end = min(self.data.len(), self.pos + literal.len());
        if &self.data[self.pos..end] == literal {
            self.pos += literal.len();
            self.skip_whitespace();

            return true;
        }

        return false;
    }

    fn expect_literal(&mut self, literal: &str) -> Result<(), ParseError> {
        if !self.match_literal(literal) {
            return Err(ParseError::new(&format!("Expected: {}", literal)));
        } else {
            return Ok(());
        }
    }

    fn expect_left_brace(&mut self) -> Result<(), ParseError> {
        return self.expect_literal("{");
    }

    fn expect_right_brace(&mut self) -> Result<(), ParseError> {
        return self.expect_literal("}");
    }

    fn match_right_brace(&mut self) -> bool {
        return self.match_literal("}");
    }

    fn error_unexpected(&mut self) -> Result<(), ParseError> {
        return Err(ParseError::new("Unexpected token"));
    }

    fn parse_float(&mut self) -> Result<f32, ParseError> {
        let c = self.data.chars().nth(self.pos).unwrap_or_default();
        if self.match_end() || !(char::is_digit(c, 10) || c == '-' || c == '.') {
            return Err(ParseError::new("Expected <float>"));
        }

        let mut pos = self.pos + 1;
        while pos < self.data.len() {
            let c = self.data.chars().nth(pos).unwrap_or_default();
            if char::is_digit(c, 10) || c == '.' || c == '-' {
                pos += 1;
            } else {
                break;
            }
        }

        let s = &self.data[self.pos..pos];
        let f = s.parse::<f32>().unwrap_or_default();

        self.pos = pos;
        self.skip_whitespace();
        return Ok(f);
    }

    fn parse_string(&mut self) -> Result<String, ParseError> {
        let c = self.data.chars().nth(self.pos).unwrap_or_default();
        if self.match_end() || c != '"' {
            return Err(ParseError::new("Expected <string>"));
        }

        let mut pos = self.pos + 1;
        loop {
            let c = self.data.chars().nth(pos).unwrap_or_default();
            if c == '"' {
                break;
            }
            pos += 1;
            if pos == self.data.len() {
                return Err(ParseError::new("Unterminated string constant"));
            }
        }

        let ret = self.data[self.pos + 1..pos].to_string();
        self.pos = pos + 1;

        self.skip_whitespace();

        return Ok(ret);
    }

    fn parse_float_triple(&mut self) -> Result<[f32; 3], ParseError> {
        let mut values = [0.0; 3];

        self.expect_literal("<")?;
        values[0] = self.parse_float()?;
        self.expect_literal(",")?;
        values[1] = self.parse_float()?;
        self.expect_literal(",")?;
        values[2] = self.parse_float()?;
        self.expect_literal(">")?;
        
        return Ok(values);
    }

    fn parse_color(&mut self) -> Result<Color, ParseError> {
        let values = self.parse_float_triple()?;
        return Ok(Color::new(values[0], values[1], values[2]));
    }

    fn parse_point(&mut self) -> Result<Point3, ParseError> {
        let values = self.parse_float_triple()?;
        return Ok(Point3::new(values[0], values[1], values[2]));
    }

    fn parse_radiance(&mut self) -> Result<Radiance, ParseError> {
        let values = self.parse_float_triple()?;
        return Ok(Radiance::new(values[0], values[1], values[2]));
    }

    fn parse_vector(&mut self) -> Result<Vec3, ParseError> {
        let values = self.parse_float_triple()?;
        return Ok(Vec3::new(values[0], values[1], values[2]));
    }

    fn parse_scene(&mut self) -> Result<Scene, ParseError> {
        let mut camera = None;
        let mut primitives = Vec::new();
        let mut lights = Vec::new();

        while !self.match_end() {
            if let Some(new_camera) = self.try_parse_camera()? {
                camera = Some(new_camera);
                continue;
            } else if let Some(primitive) = self.try_parse_primitive()? {
                primitives.push(primitive);
                continue;
            } else if let Some(light) = self.try_parse_light()? {
                lights.push(light);
                continue;
            } else {
                self.error_unexpected()?;
            }
        }

        return Ok(Scene::new(camera.expect("Camera expected"), primitives, lights));
    }

    fn try_parse_camera(&mut self) -> Result<Option<Camera>, ParseError> {
        if !self.match_literal("camera") {
            return Ok(None);
        }
        self.expect_left_brace()?;

        let position = self.parse_point()?;
        let look_at = self.parse_point()?;
        let focal_length = self.parse_float()?;
        let aperture_size = self.parse_float()?;

        self.expect_right_brace()?;

        return Ok(Some(Camera::new(position, (look_at - position).normalize(), Vec3::new(0.0, 1.0, 0.0), 60.0, focal_length, aperture_size)));
    }

    fn try_parse_light(&mut self) -> Result<Option<Box<dyn Light>>, ParseError> {
        if self.match_literal("point_light") {
            self.expect_left_brace()?;

            let position = self.parse_point()?;
            let radiance = self.parse_radiance()?;

            self.expect_right_brace()?;

            return Ok(Some(Box::new(object::light::Point::new(position, radiance))));
        } else if self.match_literal("sky") {
            self.expect_left_brace()?;

            let radiance = self.parse_radiance()?;

            self.expect_right_brace()?;

            return Ok(Some(Box::new(object::light::Sky::new(radiance))));
        } else {
            return Ok(None);
        }
    }

    fn try_parse_primitive(&mut self) -> Result<Option<Primitive>, ParseError> {
        let mut shape : Box<dyn Shape>;

        if self.match_literal("sphere") {
            self.expect_left_brace()?;

            let position = self.parse_point()?;
            let radius = self.parse_float()?;
            shape = Box::new(Sphere::new(position, radius));
        } else if self.match_literal("quad") {
            self.expect_left_brace()?;

            let position = self.parse_point()?;
            let side1 = self.parse_vector()?;
            let side2 = self.parse_vector()?;
            shape = Box::new(Quad::new(position, side1, side2));
        } else if self.match_literal("model") {
            self.expect_left_brace()?;

            let filename = self.parse_string()?;
            if filename.ends_with(".ply") {
                shape = PlyLoader::load(filename.as_str()).expect("Model expected");
            } else if filename.ends_with(".bpt") {
                shape = BptLoader::load(filename.as_str()).expect("Model expected");
            } else {
                return Err(ParseError::new(&format!("Invalid model filename {}", filename)));
            }
        } else {
            return Ok(None);
        }

        let mut surface = None;
        while !self.match_right_brace() {
            if let Some(new_surface) = self.try_parse_surface()? {
                surface = Some(new_surface);
                continue;
            } else if let Some(transformation) = self.try_parse_transformation()? {
                shape = Box::new(Transformed::new(shape, transformation));
                continue;
            } else {
                self.error_unexpected()?;
            }
        }
    
        return Ok(Some(Primitive::new(shape, surface.expect("Surface expected"))));
    }

    fn try_parse_surface(&mut self) -> Result<Option<Surface>, ParseError> {
        if !self.match_literal("surface") {
            return Ok(None);
        }
        self.expect_left_brace()?;

        let mut albedo = None;
        let mut brdfs = Vec::new();
        let mut transmit_ior = 0.0;
        let mut radiance = Radiance::ZERO;
        let mut normal_map = None;

        while !self.match_right_brace() {
            if let Some(new_albedo) = self.try_parse_albedo()? {
                albedo = Some(new_albedo);
                continue;
            } else if let Some((new_brdfs, new_transmit_ior)) = self.try_parse_brdfs()? {
                brdfs = new_brdfs;
                transmit_ior = new_transmit_ior;
                continue;
            } else if let Some(new_normal_map) = self.try_parse_normal_map()? {
                normal_map = Some(new_normal_map);
                continue;
            } else if self.match_literal("radiance") {
                radiance = self.parse_radiance()?;
                continue;
            } else {
                self.error_unexpected()?;
            }
        }

        return Ok(Some(Surface::new(albedo.expect("Albedo expected"), brdfs, transmit_ior, radiance, normal_map)));
    }

    fn try_parse_transformation(&mut self) -> Result<Option<Transformation>, ParseError> {
        if !self.match_literal("transform") {
            return Ok(None);
        }
        self.expect_left_brace()?;

        let mut result = Transformation::identity();
        while !self.match_right_brace() {
            if self.match_literal("translate") {
                let vector = self.parse_vector()?;
                result = result.transform(Transformation::translate(vector));
                continue;
            } else if self.match_literal("rotate") {
                let vector = self.parse_vector()?;
                result = result.transform(Transformation::rotate(vector));
                continue;
            } else if self.match_literal("scale") {
                let vector = self.parse_vector()?;
                result = result.transform(Transformation::scale(vector));
                continue;
            } else if self.match_literal("uniform_scale") {
                let scale = self.parse_float()?;
                result = result.transform(Transformation::uniform_scale(scale));
                continue;
            } else {
                self.error_unexpected()?;
            }
        }

        return Ok(Some(result));
    }

    fn try_parse_normal_map(&mut self) -> Result<Option<NormalMap>, ParseError> {
        if !self.match_literal("normal_map") {
            return Ok(None);
        }
        let filename = self.parse_string()?;
        let magnitude = self.parse_float()?;

        if let Ok(texture) = BmpLoader::load(filename.as_str()) {        
            return Ok(Some(NormalMap::new(texture, magnitude)));
        } else {
            return Ok(None);
        }
    }

    fn try_parse_albedo(&mut self) -> Result<Option<Box<dyn Albedo>>, ParseError> {
        if !self.match_literal("albedo") {
            return Ok(None);
        }
        self.expect_left_brace()?;

        let mut albedo: Option<Box<dyn Albedo>> = None;
        if self.match_literal("color") {
            let color = self.parse_color()?;
            albedo = Some(Box::new(Solid::new(color)));
        } else if self.match_literal("texture") {
            let filename = self.parse_string()?;
            if let Ok(texture) = BmpLoader::load(filename.as_str()) {
                albedo = Some(Box::new(Texture::new(texture)));
            }
        } else {
            self.error_unexpected()?;
        }

        self.expect_right_brace()?;
        return Ok(albedo);
    }

    fn try_parse_brdfs(&mut self) -> Result<Option<(Vec<Box<dyn Brdf>>, f32)>, ParseError> {
        if !self.match_literal("brdf") {
            return Ok(None);
        }
        self.expect_left_brace()?;

        let mut brdfs: Vec<Box<dyn Brdf>> = Vec::new();
        let mut transmit_ior = 0.0;
        while !self.match_right_brace() {
            if self.match_literal("lambert") {
                let strength = self.parse_float()?;
                brdfs.push(Box::new(Lambert::new(strength)));
                continue;
            } else if self.match_literal("phong") {
                let strength = self.parse_float()?;
                let power = self.parse_float()?;
                brdfs.push(Box::new(Phong::new(strength, power)));
                continue;
            } else if self.match_literal("oren_nayar") {
                let strength = self.parse_float()?;
                let roughness = self.parse_float()?;
                brdfs.push(Box::new(OrenNayar::new(strength, roughness)));
                continue;
            } else if self.match_literal("torrance_sparrow") {
                let strength = self.parse_float()?;
                let roughness = self.parse_float()?;
                let ior = self.parse_float()?;
                brdfs.push(Box::new(TorranceSparrow::new(strength, roughness, ior)));
                continue;
            } else if self.match_literal("transmit") {
                transmit_ior = self.parse_float()?;
                continue;
            } else {
                self.error_unexpected()?;
            }
        }

        return Ok(Some((brdfs, transmit_ior)));
    }
}

File: Parse/SceneParser.cpp
#include "Parse/SceneParser.hpp"

#include "Object/Shape/Transformed.hpp"

#include "Object/Albedo/Solid.hpp"
#include "Object/Albedo/Texture.hpp"

#include "Object/Brdf/Lambert.hpp"
#include "Object/Brdf/Phong.hpp"
#include "Object/Brdf/OrenNayar.hpp"
#include "Object/Brdf/TorranceSparrow.hpp"

#include "Parse/BptLoader.hpp"
#include "Parse/PlyLoader.hpp"
#include "Parse/BmpLoader.hpp"

#include <sstream>
#include <iostream>

namespace Parse {
    struct ParseException {
        ParseException(const std::string &m) : message(m) {}
 
        std::string message;
    };

    SceneParser::SceneParser(const std::string &filename)
    : mFile(filename, std::ios_base::binary)
    {
    }

    std::unique_ptr<Object::Scene> SceneParser::parse()
    {
        if(!mFile.good()) {
            return nullptr;
        }

        mFile.seekg(0, std::ios_base::end);
        mData.resize(mFile.tellg());
        mFile.seekg(0);
        mFile.read(&mData[0], mData.size());
        mPos = 0;
        mLine = 1;

        skipWhitespace();

        try {
            return parseScene();
        } catch(ParseException e) {
            std::cout << "Error, line " << mLine << ": " << e.message << std::endl;
            return nullptr;
        }
    }

    void SceneParser::skipWhitespace()
    {
        while(mPos < mData.size()) {
            if(std::isblank(mData[mPos]) || mData[mPos] == '\r') {
                mPos++;
                continue;
            } else if(mData[mPos] == '\n') {
                mLine++;
                mPos++;
                continue;
            } else {
                break;
            }
        }
    }

    void SceneParser::throwUnexpected()
    {
        throw ParseException("Unexpected token");
    }

    float SceneParser::parseFloat()
    {
        if(matchEnd() || !(std::isdigit(mData[mPos]) || mData[mPos] == '-' || mData[mPos] == '.')) {
            throw ParseException("Expected <float>");
        }

        int pos = mPos + 1;
        while(pos < mData.size() && (std::isdigit(mData[pos]) || mData[pos] == '.' || mData[pos] == '-')) {
            pos++;
        }

        std::string s = mData.substr(mPos, pos - mPos);
        mPos = pos;

        skipWhitespace();

        return std::atof(s.c_str());
    }

    std::string SceneParser::parseString()
    {
        if(matchEnd() || mData[mPos] != '"') {
            throw ParseException("Expected <string>");
        }

        int pos = mPos + 1;
        while(mData[pos] != '"') {
            pos++;
            if(pos == mData.size()) {
                throw ParseException("Unterminated string constant");
            }
        }

        std::string ret = mData.substr(mPos + 1, pos - mPos - 1);
        mPos = pos + 1;

        skipWhitespace();

        return ret;
    }

    bool SceneParser::matchLiteral(const std::string &literal)
    {
        if(matchEnd()) {
            return false;
        }

        if(!strncmp(mData.data() + mPos, literal.c_str(), literal.size())) {
            mPos += literal.size();
            skipWhitespace();

            return true;
        }

        return false;
    }

    void SceneParser::expectLiteral(const std::string &literal)
    {
        if(!matchLiteral(literal)) {
            std::stringstream ss;
            ss << "Expected " << literal;
            throw ParseException(ss.str());
        }
    }

    bool SceneParser::matchEnd()
    {
        return mPos == mData.size();
    }

    void SceneParser::expectLeftBrace()
    {
        expectLiteral("{");
    }

    void SceneParser::expectRightBrace()
    {
        expectLiteral("}");
    }

    bool SceneParser::matchRightBrace()
    {
        return matchLiteral("}");
    }

    void SceneParser::parseFloatTriple(float items[3])
    {
        expectLiteral("<");
        items[0] = parseFloat();
        expectLiteral(",");
        items[1] = parseFloat();
        expectLiteral(",");
        items[2] = parseFloat();
        expectLiteral(">");
    }

    Math::Point SceneParser::parsePoint()
    {
        float items[3];
        parseFloatTriple(items);
        return Math::Point(items[0], items[1], items[2]);
    }

    Math::Vector SceneParser::parseVector()
    {
        float items[3];
        parseFloatTriple(items);
        return Math::Vector(items[0], items[1], items[2]);
    }

    Math::Radiance SceneParser::parseRadiance()
    {
        float items[3];
        parseFloatTriple(items);
        return Math::Radiance(items[0], items[1], items[2]);
    }

    Math::Color SceneParser::parseColor()
    {
        float items[3];
        parseFloatTriple(items);
        return Math::Color(items[0], items[1], items[2]);
    }

    std::unique_ptr<Object::Scene> SceneParser::parseScene()
    {
        std::unique_ptr<Object::Camera> camera;
        std::vector<std::unique_ptr<Object::Primitive>> primitives;
        std::vector<std::unique_ptr<Object::Light::Base>> lights;
        Math::Radiance skyRadiance;

        while(!matchEnd()) {
            if(auto cam = tryParseCamera()) {
                camera = std::move(cam);
                continue;
            }

            if(auto primitive = tryParsePrimitive()) {
                primitives.push_back(std::move(primitive));
                continue;
            }

            if(auto light = tryParseLight()) {
                lights.push_back(std::move(light));
                continue;
            }

            if(matchLiteral("sky")) {
                expectLeftBrace();
                skyRadiance = parseRadiance();
                expectRightBrace();
                continue;
            }

            throwUnexpected();
        }

        return std::make_unique<Object::Scene>(std::move(camera), std::move(primitives), std::move(lights), skyRadiance);
    }

    std::unique_ptr<Object::Camera> SceneParser::tryParseCamera()
    {
        if(!matchLiteral("camera")) {
            return nullptr;
        }
        expectLeftBrace();

        Math::Point position = parsePoint();
        Math::Point lookAt = parsePoint();
        float focalLength = parseFloat();
        float apertureSize = parseFloat();

        expectRightBrace();

        return std::make_unique<Object::Camera>(position, (lookAt - position).normalize(), Math::Vector(0, 1, 0), 60.0f, focalLength, apertureSize);
    }

    std::unique_ptr<Object::Light::Base> SceneParser::tryParseLight()
    {
        if(!matchLiteral("point_light")) {
            return nullptr;
        }
        expectLeftBrace();

        Math::Point position = parsePoint();
        Math::Radiance radiance = parseRadiance();

        expectRightBrace();

        return std::make_unique<Object::Light::Point>(position, radiance);
    }

    std::unique_ptr<Object::Primitive> SceneParser::tryParsePrimitive()
    {
        std::unique_ptr<Object::Shape::Base> shape;
        if(matchLiteral("sphere")) {
            expectLeftBrace();

            Math::Point position = parsePoint();
            float radius = parseFloat();
            shape = std::make_unique<Object::Shape::Sphere>(position, radius);
        } else if(matchLiteral("quad")) {
            expectLeftBrace();

            Math::Point position = parsePoint();
            Math::Vector side1 = parseVector();
            Math::Vector side2 = parseVector();
            shape = std::make_unique<Object::Shape::Quad>(position, side1, side2);
        } else if(matchLiteral("model")) {
            expectLeftBrace();
            
            std::string filename = parseString();
            std::string extension = filename.substr(filename.find_last_of('.'));
            
            if (extension == ".bpt") {
                shape = BptLoader::load(filename);
            } else if (extension == ".ply") {
                shape = PlyLoader::load(filename);
            } else {
                std::stringstream ss;
                ss << "Unknown model extension " << extension;
                throw ParseException(ss.str());
            }
        } else {
            return nullptr;
        }

        std::unique_ptr<Object::Surface> surface;
        Math::Transformation transformation;
        while(!matchRightBrace()) {
            if(auto newSurface = tryParseSurface()) {
                surface = std::move(newSurface);
                continue;
            }

            if(tryParseTransformation(transformation)) {
                shape = std::make_unique<Object::Shape::Transformed>(std::move(shape), transformation);
                continue;
            }

            throwUnexpected();
        }

        return std::make_unique<Object::Primitive>(std::move(shape), std::move(surface));
    }

    std::unique_ptr<Object::Surface> SceneParser::tryParseSurface()
    {
        if(!matchLiteral("surface")) {
            return nullptr;
        }
        expectLeftBrace();
        
        std::unique_ptr<Object::Albedo::Base> albedo;
        std::vector<std::unique_ptr<Object::Brdf::Base>> brdfs;
        float transmitIor = 0.0f;
        Math::Radiance radiance;
        std::unique_ptr<Object::NormalMap> normalMap;

        while(!matchRightBrace()) {
            if(auto newAlbedo = tryParseAlbedo()) {
                albedo = std::move(newAlbedo);
                continue;
            }

            if(tryParseBrdfs(brdfs, transmitIor)) {
                continue;
            }

            if(auto newNormalMap = tryParseNormalMap()) {
                normalMap = std::move(newNormalMap);
                continue;
            }

            if(matchLiteral("radiance")) {
                radiance = parseRadiance();
                continue;
            }

            throwUnexpected();
        }
        
        return std::make_unique<Object::Surface>(std::move(albedo), std::move(brdfs), transmitIor, radiance, std::move(normalMap));
    }

    bool SceneParser::tryParseTransformation(Math::Transformation &transformation)
    {
        if(!matchLiteral("transform")) {
            return false;
        }
        expectLeftBrace();

        while(!matchRightBrace()) {
            if(matchLiteral("translate")) {
                Math::Vector vector = parseVector();
                transformation = Math::Transformation::translate(vector) * transformation;
                continue;
            }

            if(matchLiteral("rotate")) {
                Math::Vector vector = parseVector();
                transformation = Math::Transformation::rotate(vector) * transformation;
                continue;
            }

            if(matchLiteral("scale")) {
                Math::Vector vector = parseVector();
                transformation = Math::Transformation::scale(vector) * transformation;
                continue;
            }

            if(matchLiteral("uniform_scale")) {
                float scale = parseFloat();
                transformation = Math::Transformation::uniformScale(scale) * transformation;
                continue;
            }

            throwUnexpected();
        }

        return true;
    }

    std::unique_ptr<Object::Albedo::Base> SceneParser::tryParseAlbedo()
    {
        if(!matchLiteral("albedo")) {
            return nullptr;
        }
        expectLeftBrace();

        std::unique_ptr<Object::Albedo::Base> albedo;

        if(matchLiteral("color")) {
            Math::Color color = parseColor();
            albedo = std::make_unique<Object::Albedo::Solid>(color);
        } else if(matchLiteral("texture")) {
            auto texture = BmpLoader::load(parseString());
            albedo = std::make_unique<Object::Albedo::Texture>(std::move(texture));
        } else {
            throwUnexpected();
        }

        expectRightBrace();

        return albedo;
    }

    bool SceneParser::tryParseBrdfs(std::vector<std::unique_ptr<Object::Brdf::Base>> &brdfs, float &transmitIor)
    {
        if(!matchLiteral("brdf")) {
            return false;
        }

        expectLeftBrace();

        while(!matchRightBrace()) {
            if(matchLiteral("lambert")) {
                float strength = parseFloat();
                brdfs.push_back(std::make_unique<Object::Brdf::Lambert>(strength));
                continue;
            }
    
            if(matchLiteral("phong")) {
                float strength = parseFloat();
                float power = parseFloat();
                brdfs.push_back(std::make_unique<Object::Brdf::Phong>(strength, power));
                continue;
            }

            if(matchLiteral("oren_nayar")) {
                float strength = parseFloat();
                float roughness = parseFloat();
                brdfs.push_back(std::make_unique<Object::Brdf::OrenNayar>(strength, roughness));
                continue;
            }
    
            if(matchLiteral("torrance_sparrow")) {
                float strength = parseFloat();
                float roughness = parseFloat();
                float ior = parseFloat();
                brdfs.push_back(std::make_unique<Object::Brdf::TorranceSparrow>(strength, roughness, ior));
                continue;
            }

            if(matchLiteral("transmit")) {
                transmitIor = parseFloat();
                continue;
            }

            throwUnexpected();
        }

        return true;
    }

    std::unique_ptr<Object::NormalMap> SceneParser::tryParseNormalMap()
    {
        if(!matchLiteral("normal_map")) {
            return nullptr;
        }
       
        auto texture = BmpLoader::load(parseString());
        float magnitude = parseFloat();

        return std::make_unique<Object::NormalMap>(std::move(texture), magnitude);
    }
}